src/runtime/disco/session.cc

```diff
@@ -0,0 +1,165 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+#include "./session.h"
+
+#include <tvm/runtime/packed_func.h>
+#include <tvm/runtime/registry.h>
+
+#include "./worker.h"
+
+namespace tvm {
+namespace runtime {
+
+TVM_REGISTER_OBJECT_TYPE(DRefObj);
+TVM_REGISTER_OBJECT_TYPE(CommandQueueObj);
+TVM_REGISTER_OBJECT_TYPE(SessionObj);
+
+DRefObj::~DRefObj() { Downcast<Session>(this->session)->DeallocReg(reg_id); }
+
+DRef::DRef(int reg_id, Session session) {
+  ObjectPtr<DRefObj> p = make_object<DRefObj>();
+  p->reg_id = reg_id;
+  p->session = session;
+  data_ = std::move(p);
+}
+
+TVMRetValue DRefObj::DebugGetFromRemote(int worker_id) {
+  return Downcast<Session>(this->session)->DebugGetFromRemote(this->reg_id, worker_id);
+}
+
+void CommandQueueObj::SyncWorker(int worker_id) {
+  this->BroadcastUnpacked(DiscoAction::kSyncWorker, worker_id);
+  TVMArgs args = this->RecvReplyPacked(worker_id);
+  ICHECK_EQ(args.size(), 2);
+  DiscoAction action = static_cast<DiscoAction>(args[0].operator int());
+  int ret_worker_id = args[1];
+  ICHECK(action == DiscoAction::kSyncWorker);
+  ICHECK_EQ(ret_worker_id, worker_id);
+}
+
+int SessionObj::AllocateReg() {
+  if (this->free_regs_.empty()) {
+    return this->reg_count_++;
+  }
+  int reg_id = this->free_regs_.back();
+  this->free_regs_.pop_back();
+  return reg_id;
+}
+
+void SessionObj::DeallocReg(int reg_id) {
+  cmd_->BroadcastUnpacked(DiscoAction::kKillReg, reg_id);
+  this->free_regs_.push_back(reg_id);
+}
+
+DRef SessionObj::CallPackedWithPacked(const TVMArgs& args) {
+  constexpr uint64_t one = 1;
+  constexpr int offset = 4;
+  TVMValue* values = const_cast<TVMValue*>(args.values);
+  int* type_codes = const_cast<int*>(args.type_codes);
+  int num_args = args.num_args;
+  int reg_id = AllocateReg();
+  uint64_t is_dref = 0;
```

Comment:
introduce a special type code for dref

Reply:
done. we will split it into a separate PR to main